pkg/services/ngalert/state/manager.go

```diff
@@ -199,12 +199,13 @@ func (st *Manager) ResetStateByRuleUID(ctx context.Context, rule *ngModels.Alert
 	for _, s := range states {
 		oldState := s.State
 		oldReason := s.StateReason
+		state := *s
 		now := time.Now()
-		s.SetNormal(reason, s.StartsAt, now)
-		s.LastEvaluationTime = now
-		s.Values = map[string]float64{}
+		state.SetNormal(reason, s.StartsAt, now)
+		state.LastEvaluationTime = now
+		state.Values = map[string]float64{}
 		transitions = append(transitions, StateTransition{
-			State:               s,
```

Comment:
why do we need to do that?

Reply:
This was resolved in a slack private channel. Let me explain here.

Before all these changes were added, we didn't care about the mutability of the state instances as we only needed them to send a resolved message for the ones that needed it (ie firing alerts) when we cleared the state from the cache and DB.

Now we added a previous step before sending the resolved messages (save state history `Normal (Paused`).

What is the problem? 

In this added (save state history) step we are mutating the state instances to be able to save the history as `Normal (Paused)`, this means that in the second step (sending resolved messages), we are not going to detect any instances that need to be resolved as we mutated them to `Normal`.

That's why we create a copy of the state.

We are gonna leave it for now as it is and change it in a follow-up PR.